### PR TITLE
session: support tso when use tidb_snapshot

### PR DESCRIPTION
**Pull request by zhexuany**
After this PR, `set @@tidb_snapshot= '400036290571534337'` can be used.

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
We need to update the related document.

File: types/time.go [hunk 1]
```diff
@@ -45,6 +45,10 @@ const (
 	TimeFormat = "2006-01-02 15:04:05"
 	// TimeFSPFormat is time format with fractional seconds precision.
 	TimeFSPFormat = "2006-01-02 15:04:05.000000"
+
+	// used for encodinh/decoding tso
+	physicalShiftBits = 18
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply types/time.go]: DONE

**Comment by zhexuany** [comment]
@shenli PTAL

File: sessionctx/variable/varsutil.go [hunk 2]
```diff
@@ -175,10 +175,18 @@ func setSnapshotTS(s *SessionVars, sVal string) error {
 		s.SnapshotTS = 0
 		return nil
 	}
+
+	if tso, err2 := strconv.ParseUint(sVal, 10, 64); err2 == nil {
+		s.SnapshotTS = tso
+		return errors.Trace(nil)
```
Comment (shenli) [inline sessionctx/variable/varsutil.go]: return nil

**Comment by zhexuany** [comment]
@shenli PTAL

File: store/tikv/oracle/oracle.go [hunk 3]
```diff
@@ -48,3 +48,15 @@ func ExtractPhysical(ts uint64) int64 {
 func GetPhysical(t time.Time) int64 {
 	return t.UnixNano() / int64(time.Millisecond)
 }
+
+// EncodeTso encodes a millisecond into tso.
+func EncodeTso(ts int64) uint64 {
```
Comment (zimulala) [inline store/tikv/oracle/oracle.go]: s/Tso/TSO ?

File: store/tikv/oracle/oracle.go [hunk 4]
```diff
@@ -48,3 +48,15 @@ func ExtractPhysical(ts uint64) int64 {
...
+	return uint64(ts) << physicalShiftBits
+}
+
+// DecodeTso decodes a tso into a golang's time struct.
+func DecodeTso(tso uint64) time.Time {
```
Comment (zimulala) [inline store/tikv/oracle/oracle.go]: Ditto.

File: sessionctx/variable/varsutil.go [hunk 5]
```diff
@@ -175,10 +175,18 @@ func setSnapshotTS(s *SessionVars, sVal string) error {
...
 		return nil
 	}
+
+	if tso, err2 := strconv.ParseUint(sVal, 10, 64); err2 == nil {
+		s.SnapshotTS = tso
+		return nil
+	}
+
 	t, err := types.ParseTime(s.StmtCtx, sVal, mysql.TypeTimestamp, types.MaxFsp)
+
```
Comment (zimulala) [inline sessionctx/variable/varsutil.go]: Remove this useless line.

File: sessionctx/variable/varsutil.go [hunk 6]
```diff
@@ -175,10 +175,18 @@ func setSnapshotTS(s *SessionVars, sVal string) error {
 		s.SnapshotTS = 0
 		return nil
 	}
+
+	if tso, err2 := strconv.ParseUint(sVal, 10, 64); err2 == nil {
```
Comment (zimulala) [inline sessionctx/variable/varsutil.go]: s/err2/err ?

File: store/tikv/oracle/oracle.go (hunk 4 above)
Comment (zimulala) [reply store/tikv/oracle/oracle.go]: Do we use this function? Could we remove it？

**Comment by zhexuany** [comment]
@zimulala All comments are addressed. PTAL

File: store/tikv/oracle/oracle.go [hunk 7]
```diff
@@ -48,3 +48,15 @@ func ExtractPhysical(ts uint64) int64 {
...
 	return t.UnixNano() / int64(time.Millisecond)
 }
+
+// EncodeTSO encodes a millisecond into tso.
+func EncodeTSO(ts int64) uint64 {
+	return uint64(ts) << physicalShiftBits
+}
+
+// DecodeTSO decodes a tso into a golang's time struct.
+func DecodeTSO(tso uint64) time.Time {
```
Comment (zimulala) [inline store/tikv/oracle/oracle.go]: Do we use this function? Could we remove it？

**Comment by shenli** [comment]
@zhexuany We need to cherry-pick this to the release-2.0 branch.

**Comment by zhexuany** [comment]
confirmed with @WangXiangUSTC, as long as this PR be in part of 2.0.4 release, they are fine with this.

**Pull request merged**